_posts/2018-01-07-building-serverless-radiator.md

```diff
@@ -0,0 +1,225 @@
+- Serverless 
+---
+
+***There exists many commercial products for monitoring development pipelines and deployments. What could be more fun than make your own?***
+
+**Modern** software development environment is most likely composed of multiple pipelines. There can be pipelines for build artifacts, infrastructure building etc. It is vital to notice fast when the builds fail. Monitoring only the pipelines is not necessary enough. Also deployments should be monitored for noticing potential problems in performance etc. 
+
+One project might have different accounts for development, testing/integration and production. This means that monitoring simply one account is not even closely enough. For example in my current project at Solita, we have 3 different accounts which have 7 pipelines, over 60 CloudWatch alarms and some exposed metrics from deployments.
+
+One good solution is to have a radiator which shows development status:
+
+![Alarm!](/img/building-serverless-radiator/complete-radiator.png)
+
+Because making your own is interesting and fun I chose to make my own radiator. For cost optimization I chose serverless approach with AWS [Lambdas](https://aws.amazon.com/lambda/). Lets start our journey from how to expose pipelines, alarms and metrics from AWS accounts.
+
+## Exposing environment statuses from AWS accounts
+
+AWS [CodePipeline](https://aws.amazon.com/codepipeline/) is an AWS software product for building, testing and installing software to the AWS environment. It integrates very nicely with AWS [CloudFormation](https://aws.amazon.com/cloudformation/) and is a pay only what you use product. There is no need to provision EC2 virtual machines for builds - AWS will provision them for you.
+
+One can get statuses from CodePipeline by using AWS CLI or AWS SDK which is available for many different languages. AWS [CloudWatch](https://aws.amazon.com/cloudwatch/) alarms and metrics are also accessible via CLI and SDK. 
+
+One way to extract data would be to make an AWS user to every account, generate a access key id and a secret key and poll end points directly from the radiator software. A variant of this would be to make a AWS role and configure cross account permissions to switch to that role. 
+
+Third way which I chose is to make a Lambda function which exposes statuses in one JSON data response. That way it is possible to get all necessary data from one endpoint. The endpoint should be protected with an api key which is supported quite nicely in the Amazon [API Gateway](https://aws.amazon.com/api-gateway/). The radiator should poll the endpoints and generate a view of the account status.
+
+AWS Lambda environment currently supports Node.js, Java, C# and Python languages. For an endpoint exposer I chose Python which have excellent [Boto 3](https://boto3.readthedocs.io/en/latest/) SDK for accessing AWS.
+
+First I started from what I want as a result. The result should be JSON response with different keys for alarms, pipelines and metrics. 
```

Comment:
"I started by defining what I wanted to get as a result."

"The result should be *a* JSON response…"